fsspec/spec.py

```diff
@@ -803,13 +803,13 @@ def put_file(self, lpath, rpath, callback=_DEFAULT_CALLBACK, **kwargs):
             return None
 
         with open(lpath, "rb") as f1:
-            callback.set_size(f1.seek(0, 2))
+            size = f1.seek(0, 2)
+            callback.set_size(size)
             f1.seek(0)
 
             self.mkdirs(self._parent(os.fspath(rpath)), exist_ok=True)
             with self.open(rpath, "wb", **kwargs) as f2:
-                data = True
-                while data:
+                while f2.tell() < size:
```

Comment:
why not `f1.tell() < size`? If `f2.write()` silently failed to write a chunk, then this would go into an infinite loop where `f1.read()` thinks it's done but `f2` never gets the whole file

Reply:
Hm, that really shouldn't happen, but I'd be OK with your suggestion. Do you want to make that follow-on PR?